{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "04732d81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pyfaidx in /mnt/raid1/thalassini/home/anaconda3/lib/python3.11/site-packages (0.7.2.2)\n",
      "Requirement already satisfied: six in /mnt/raid1/thalassini/home/anaconda3/lib/python3.11/site-packages (from pyfaidx) (1.16.0)\n",
      "Requirement already satisfied: setuptools in /mnt/raid1/thalassini/home/anaconda3/lib/python3.11/site-packages (from pyfaidx) (68.0.0)\n",
      "Requirement already satisfied: importlib-metadata in /mnt/raid1/thalassini/home/anaconda3/lib/python3.11/site-packages (from pyfaidx) (6.0.0)\n",
      "Requirement already satisfied: zipp>=0.5 in /mnt/raid1/thalassini/home/anaconda3/lib/python3.11/site-packages (from importlib-metadata->pyfaidx) (3.11.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install  pyfaidx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "c5272264",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting torch\n",
      "  Obtaining dependency information for torch from https://files.pythonhosted.org/packages/da/6a/7fb9d82db4568834ff6d4df2fe3b143de4ed65a3f8f93e7daed703626cb6/torch-2.1.2-cp311-cp311-manylinux1_x86_64.whl.metadata\n",
      "  Downloading torch-2.1.2-cp311-cp311-manylinux1_x86_64.whl.metadata (25 kB)\n",
      "Requirement already satisfied: filelock in /mnt/raid1/thalassini/home/anaconda3/lib/python3.11/site-packages (from torch) (3.9.0)\n",
      "Requirement already satisfied: typing-extensions in /mnt/raid1/thalassini/home/anaconda3/lib/python3.11/site-packages (from torch) (4.7.1)\n",
      "Requirement already satisfied: sympy in /mnt/raid1/thalassini/home/anaconda3/lib/python3.11/site-packages (from torch) (1.11.1)\n",
      "Requirement already satisfied: networkx in /mnt/raid1/thalassini/home/anaconda3/lib/python3.11/site-packages (from torch) (3.1)\n",
      "Requirement already satisfied: jinja2 in /mnt/raid1/thalassini/home/anaconda3/lib/python3.11/site-packages (from torch) (3.1.2)\n",
      "Requirement already satisfied: fsspec in /mnt/raid1/thalassini/home/anaconda3/lib/python3.11/site-packages (from torch) (2023.4.0)\n",
      "Collecting nvidia-cuda-nvrtc-cu12==12.1.105 (from torch)\n",
      "  Downloading nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)\n",
      "\u001b[2K     \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m23.7/23.7 MB\u001b[0m \u001b[31m31.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0mm eta \u001b[36m0:00:01\u001b[0m[36m0:00:01\u001b[0m\n",
      "\u001b[?25hCollecting nvidia-cuda-runtime-cu12==12.1.105 (from torch)\n",
      "  Downloading nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (823 kB)\n",
      "\u001b[2K     \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m823.6/823.6 kB\u001b[0m \u001b[31m73.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hCollecting nvidia-cuda-cupti-cu12==12.1.105 (from torch)\n",
      "  Downloading nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (14.1 MB)\n",
      "\u001b[2K     \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m14.1/14.1 MB\u001b[0m \u001b[31m38.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0mm eta \u001b[36m0:00:01\u001b[0m0:01\u001b[0m:01\u001b[0m\n",
      "\u001b[?25hCollecting nvidia-cudnn-cu12==8.9.2.26 (from torch)\n",
      "  Obtaining dependency information for nvidia-cudnn-cu12==8.9.2.26 from https://files.pythonhosted.org/packages/ff/74/a2e2be7fb83aaedec84f391f082cf765dfb635e7caa9b49065f73e4835d8/nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl.metadata\n",
      "  Downloading nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)\n",
      "Collecting nvidia-cublas-cu12==12.1.3.1 (from torch)\n",
      "  Downloading nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl (410.6 MB)\n",
      "\u001b[2K     \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m410.6/410.6 MB\u001b[0m \u001b[31m5.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0mm eta \u001b[36m0:00:01\u001b[0m[36m0:00:01\u001b[0m\n",
      "\u001b[?25hCollecting nvidia-cufft-cu12==11.0.2.54 (from torch)\n",
      "  Downloading nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl (121.6 MB)\n",
      "\u001b[2K     \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m121.6/121.6 MB\u001b[0m \u001b[31m16.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0mm eta \u001b[36m0:00:01\u001b[0m[36m0:00:01\u001b[0m\n",
      "\u001b[?25hCollecting nvidia-curand-cu12==10.3.2.106 (from torch)\n",
      "  Downloading nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl (56.5 MB)\n",
      "\u001b[2K     \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m56.5/56.5 MB\u001b[0m \u001b[31m25.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0mm eta \u001b[36m0:00:01\u001b[0m0:01\u001b[0m:01\u001b[0m\n",
      "\u001b[?25hCollecting nvidia-cusolver-cu12==11.4.5.107 (from torch)\n",
      "  Downloading nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl (124.2 MB)\n",
      "\u001b[2K     \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m124.2/124.2 MB\u001b[0m \u001b[31m12.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0mm eta \u001b[36m0:00:01\u001b[0m[36m0:00:01\u001b[0m\n",
      "\u001b[?25hCollecting nvidia-cusparse-cu12==12.1.0.106 (from torch)\n",
      "  Downloading nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl (196.0 MB)\n",
      "\u001b[2K     \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m196.0/196.0 MB\u001b[0m \u001b[31m10.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0mm eta \u001b[36m0:00:01\u001b[0m[36m0:00:01\u001b[0m\n",
      "\u001b[?25hCollecting nvidia-nccl-cu12==2.18.1 (from torch)\n",
      "  Downloading nvidia_nccl_cu12-2.18.1-py3-none-manylinux1_x86_64.whl (209.8 MB)\n",
      "\u001b[2K     \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m209.8/209.8 MB\u001b[0m \u001b[31m9.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0mm eta \u001b[36m0:00:01\u001b[0m[36m0:00:01\u001b[0m\n",
      "\u001b[?25hCollecting nvidia-nvtx-cu12==12.1.105 (from torch)\n",
      "  Downloading nvidia_nvtx_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (99 kB)\n",
      "\u001b[2K     \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m99.1/99.1 kB\u001b[0m \u001b[31m31.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hCollecting triton==2.1.0 (from torch)\n",
      "  Obtaining dependency information for triton==2.1.0 from https://files.pythonhosted.org/packages/5c/c1/54fffb2eb13d293d9a429fead3646752ea190de0229bcf3d591ba2481263/triton-2.1.0-0-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl.metadata\n",
      "  Downloading triton-2.1.0-0-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl.metadata (1.3 kB)\n",
      "Collecting nvidia-nvjitlink-cu12 (from nvidia-cusolver-cu12==11.4.5.107->torch)\n",
      "  Obtaining dependency information for nvidia-nvjitlink-cu12 from https://files.pythonhosted.org/packages/1e/07/bf730d44c2fe1b676ad9cc2be5f5f861eb5d153fb6951987a2d6a96379a9/nvidia_nvjitlink_cu12-12.3.101-py3-none-manylinux1_x86_64.whl.metadata\n",
      "  Downloading nvidia_nvjitlink_cu12-12.3.101-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /mnt/raid1/thalassini/home/anaconda3/lib/python3.11/site-packages (from jinja2->torch) (2.1.1)\n",
      "Requirement already satisfied: mpmath>=0.19 in /mnt/raid1/thalassini/home/anaconda3/lib/python3.11/site-packages (from sympy->torch) (1.3.0)\n",
      "Downloading torch-2.1.2-cp311-cp311-manylinux1_x86_64.whl (670.2 MB)\n",
      "\u001b[2K   \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m670.2/670.2 MB\u001b[0m \u001b[31m3.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m0m eta \u001b[36m0:00:01\u001b[0m[36m0:00:01\u001b[0m\n",
      "\u001b[?25hDownloading nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl (731.7 MB)\n",
      "\u001b[2K   \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m731.7/731.7 MB\u001b[0m \u001b[31m3.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m0m eta \u001b[36m0:00:01\u001b[0m[36m0:00:01\u001b[0m\n",
      "\u001b[?25hDownloading triton-2.1.0-0-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (89.2 MB)\n",
      "\u001b[2K   \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m89.2/89.2 MB\u001b[0m \u001b[31m16.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m0m eta \u001b[36m0:00:01\u001b[0m0:01\u001b[0m:01\u001b[0m\n",
      "\u001b[?25hDownloading nvidia_nvjitlink_cu12-12.3.101-py3-none-manylinux1_x86_64.whl (20.5 MB)\n",
      "\u001b[2K   \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m20.5/20.5 MB\u001b[0m \u001b[31m56.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m0m eta \u001b[36m0:00:01\u001b[0m0:01\u001b[0m:01\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: triton, nvidia-nvtx-cu12, nvidia-nvjitlink-cu12, nvidia-nccl-cu12, nvidia-curand-cu12, nvidia-cufft-cu12, nvidia-cuda-runtime-cu12, nvidia-cuda-nvrtc-cu12, nvidia-cuda-cupti-cu12, nvidia-cublas-cu12, nvidia-cusparse-cu12, nvidia-cudnn-cu12, nvidia-cusolver-cu12, torch\n",
      "Successfully installed nvidia-cublas-cu12-12.1.3.1 nvidia-cuda-cupti-cu12-12.1.105 nvidia-cuda-nvrtc-cu12-12.1.105 nvidia-cuda-runtime-cu12-12.1.105 nvidia-cudnn-cu12-8.9.2.26 nvidia-cufft-cu12-11.0.2.54 nvidia-curand-cu12-10.3.2.106 nvidia-cusolver-cu12-11.4.5.107 nvidia-cusparse-cu12-12.1.0.106 nvidia-nccl-cu12-2.18.1 nvidia-nvjitlink-cu12-12.3.101 nvidia-nvtx-cu12-12.1.105 torch-2.1.2 triton-2.1.0\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install torch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "f344a7f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c44c44b3",
   "metadata": {},
   "source": [
    "# Convert fimo.tsv to bed format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "bb5e6e15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif_id</th>\n",
       "      <th>motif_alt_id</th>\n",
       "      <th>sequence_name</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>p-value</th>\n",
       "      <th>q-value</th>\n",
       "      <th>matched_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr10</td>\n",
       "      <td>11458.0</td>\n",
       "      <td>11472.0</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>2.930000e-10</td>\n",
       "      <td>0.026</td>\n",
       "      <td>GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr12</td>\n",
       "      <td>547112.0</td>\n",
       "      <td>547126.0</td>\n",
       "      <td>-</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>2.930000e-10</td>\n",
       "      <td>0.026</td>\n",
       "      <td>GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr1</td>\n",
       "      <td>869920.0</td>\n",
       "      <td>869934.0</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>2.930000e-10</td>\n",
       "      <td>0.026</td>\n",
       "      <td>GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr1</td>\n",
       "      <td>904770.0</td>\n",
       "      <td>904784.0</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>2.930000e-10</td>\n",
       "      <td>0.026</td>\n",
       "      <td>GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr6</td>\n",
       "      <td>1397165.0</td>\n",
       "      <td>1397179.0</td>\n",
       "      <td>-</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>2.930000e-10</td>\n",
       "      <td>0.026</td>\n",
       "      <td>GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480300</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr1</td>\n",
       "      <td>245200915.0</td>\n",
       "      <td>245200929.0</td>\n",
       "      <td>-</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>4.490000e-05</td>\n",
       "      <td>0.548</td>\n",
       "      <td>ATCAACAGGGGTCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480301</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr1</td>\n",
       "      <td>245560998.0</td>\n",
       "      <td>245561012.0</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>4.490000e-05</td>\n",
       "      <td>0.548</td>\n",
       "      <td>AACAGAAGGTGGCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480302</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr1</td>\n",
       "      <td>246738775.0</td>\n",
       "      <td>246738789.0</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>4.490000e-05</td>\n",
       "      <td>0.548</td>\n",
       "      <td>CCCGCGAGGTGGAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480303</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr1</td>\n",
       "      <td>247423982.0</td>\n",
       "      <td>247423996.0</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>4.490000e-05</td>\n",
       "      <td>0.548</td>\n",
       "      <td>AGCAGGAGAGGGAGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480304</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr1</td>\n",
       "      <td>247470114.0</td>\n",
       "      <td>247470128.0</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>4.490000e-05</td>\n",
       "      <td>0.548</td>\n",
       "      <td>ATCACAAGAGGTCAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>480305 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        motif_id motif_alt_id sequence_name        start         stop strand  \\\n",
       "0       MA0139.2         CTCF         chr10      11458.0      11472.0      +   \n",
       "1       MA0139.2         CTCF         chr12     547112.0     547126.0      -   \n",
       "2       MA0139.2         CTCF          chr1     869920.0     869934.0      +   \n",
       "3       MA0139.2         CTCF          chr1     904770.0     904784.0      +   \n",
       "4       MA0139.2         CTCF          chr6    1397165.0    1397179.0      -   \n",
       "...          ...          ...           ...          ...          ...    ...   \n",
       "480300  MA0139.2         CTCF          chr1  245200915.0  245200929.0      -   \n",
       "480301  MA0139.2         CTCF          chr1  245560998.0  245561012.0      +   \n",
       "480302  MA0139.2         CTCF          chr1  246738775.0  246738789.0      +   \n",
       "480303  MA0139.2         CTCF          chr1  247423982.0  247423996.0      +   \n",
       "480304  MA0139.2         CTCF          chr1  247470114.0  247470128.0      +   \n",
       "\n",
       "          score       p-value  q-value matched_sequence  \n",
       "0       24.2951  2.930000e-10    0.026  GCCACCAGGGGGCGC  \n",
       "1       24.2951  2.930000e-10    0.026  GCCACCAGGGGGCGC  \n",
       "2       24.2951  2.930000e-10    0.026  GCCACCAGGGGGCGC  \n",
       "3       24.2951  2.930000e-10    0.026  GCCACCAGGGGGCGC  \n",
       "4       24.2951  2.930000e-10    0.026  GCCACCAGGGGGCGC  \n",
       "...         ...           ...      ...              ...  \n",
       "480300  10.0984  4.490000e-05    0.548  ATCAACAGGGGTCAC  \n",
       "480301  10.0984  4.490000e-05    0.548  AACAGAAGGTGGCCC  \n",
       "480302  10.0984  4.490000e-05    0.548  CCCGCGAGGTGGAGG  \n",
       "480303  10.0984  4.490000e-05    0.548  AGCAGGAGAGGGAGC  \n",
       "480304  10.0984  4.490000e-05    0.548  ATCACAAGAGGTCAG  \n",
       "\n",
       "[480305 rows x 10 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the TSV file into a DataFrame\n",
    "#file_path=input(\"Give fimo.tsv file path: \")\n",
    "file_path = '/mnt/raid1/thalassini/home/fimo/encode_grch38_max/fimo.tsv'\n",
    "df = pd.read_csv(file_path, sep='\\t')\n",
    "df.dropna(inplace=True)\n",
    "\n",
    "# Display the DataFrame\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f595c1f3",
   "metadata": {},
   "source": [
    "## Rename columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "4b9d8f69",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif_id</th>\n",
       "      <th>motif_alt_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>p-value</th>\n",
       "      <th>q-value</th>\n",
       "      <th>matched_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr10</td>\n",
       "      <td>11458.0</td>\n",
       "      <td>11472.0</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>2.930000e-10</td>\n",
       "      <td>0.026</td>\n",
       "      <td>GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr12</td>\n",
       "      <td>547112.0</td>\n",
       "      <td>547126.0</td>\n",
       "      <td>-</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>2.930000e-10</td>\n",
       "      <td>0.026</td>\n",
       "      <td>GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr1</td>\n",
       "      <td>869920.0</td>\n",
       "      <td>869934.0</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>2.930000e-10</td>\n",
       "      <td>0.026</td>\n",
       "      <td>GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr1</td>\n",
       "      <td>904770.0</td>\n",
       "      <td>904784.0</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>2.930000e-10</td>\n",
       "      <td>0.026</td>\n",
       "      <td>GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr6</td>\n",
       "      <td>1397165.0</td>\n",
       "      <td>1397179.0</td>\n",
       "      <td>-</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>2.930000e-10</td>\n",
       "      <td>0.026</td>\n",
       "      <td>GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480300</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr1</td>\n",
       "      <td>245200915.0</td>\n",
       "      <td>245200929.0</td>\n",
       "      <td>-</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>4.490000e-05</td>\n",
       "      <td>0.548</td>\n",
       "      <td>ATCAACAGGGGTCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480301</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr1</td>\n",
       "      <td>245560998.0</td>\n",
       "      <td>245561012.0</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>4.490000e-05</td>\n",
       "      <td>0.548</td>\n",
       "      <td>AACAGAAGGTGGCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480302</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr1</td>\n",
       "      <td>246738775.0</td>\n",
       "      <td>246738789.0</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>4.490000e-05</td>\n",
       "      <td>0.548</td>\n",
       "      <td>CCCGCGAGGTGGAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480303</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr1</td>\n",
       "      <td>247423982.0</td>\n",
       "      <td>247423996.0</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>4.490000e-05</td>\n",
       "      <td>0.548</td>\n",
       "      <td>AGCAGGAGAGGGAGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480304</th>\n",
       "      <td>MA0139.2</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>chr1</td>\n",
       "      <td>247470114.0</td>\n",
       "      <td>247470128.0</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>4.490000e-05</td>\n",
       "      <td>0.548</td>\n",
       "      <td>ATCACAAGAGGTCAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>480305 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        motif_id motif_alt_id Chromosome        Start          End strand  \\\n",
       "0       MA0139.2         CTCF      chr10      11458.0      11472.0      +   \n",
       "1       MA0139.2         CTCF      chr12     547112.0     547126.0      -   \n",
       "2       MA0139.2         CTCF       chr1     869920.0     869934.0      +   \n",
       "3       MA0139.2         CTCF       chr1     904770.0     904784.0      +   \n",
       "4       MA0139.2         CTCF       chr6    1397165.0    1397179.0      -   \n",
       "...          ...          ...        ...          ...          ...    ...   \n",
       "480300  MA0139.2         CTCF       chr1  245200915.0  245200929.0      -   \n",
       "480301  MA0139.2         CTCF       chr1  245560998.0  245561012.0      +   \n",
       "480302  MA0139.2         CTCF       chr1  246738775.0  246738789.0      +   \n",
       "480303  MA0139.2         CTCF       chr1  247423982.0  247423996.0      +   \n",
       "480304  MA0139.2         CTCF       chr1  247470114.0  247470128.0      +   \n",
       "\n",
       "          score       p-value  q-value matched_sequence  \n",
       "0       24.2951  2.930000e-10    0.026  GCCACCAGGGGGCGC  \n",
       "1       24.2951  2.930000e-10    0.026  GCCACCAGGGGGCGC  \n",
       "2       24.2951  2.930000e-10    0.026  GCCACCAGGGGGCGC  \n",
       "3       24.2951  2.930000e-10    0.026  GCCACCAGGGGGCGC  \n",
       "4       24.2951  2.930000e-10    0.026  GCCACCAGGGGGCGC  \n",
       "...         ...           ...      ...              ...  \n",
       "480300  10.0984  4.490000e-05    0.548  ATCAACAGGGGTCAC  \n",
       "480301  10.0984  4.490000e-05    0.548  AACAGAAGGTGGCCC  \n",
       "480302  10.0984  4.490000e-05    0.548  CCCGCGAGGTGGAGG  \n",
       "480303  10.0984  4.490000e-05    0.548  AGCAGGAGAGGGAGC  \n",
       "480304  10.0984  4.490000e-05    0.548  ATCACAAGAGGTCAG  \n",
       "\n",
       "[480305 rows x 10 columns]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.rename(columns={\"sequence_name\": \"Chromosome\", \"start\": \"Start\", \"stop\": \"End\"}, inplace=True)\n",
    "df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14197c79",
   "metadata": {},
   "source": [
    "## Convert coordinated to integers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "1d2a4466",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"Start\"]=df[\"Start\"].astype('int')\n",
    "df[\"End\"]=df[\"End\"].astype('int')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb98692e",
   "metadata": {},
   "source": [
    "## Merge \"p-value, q-value & matched_sequence\" in column \"name\" "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "8851a0d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>11458</td>\n",
       "      <td>11472</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>547112</td>\n",
       "      <td>547126</td>\n",
       "      <td>-</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>869920</td>\n",
       "      <td>869934</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>904770</td>\n",
       "      <td>904784</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>1397165</td>\n",
       "      <td>1397179</td>\n",
       "      <td>-</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480300</th>\n",
       "      <td>chr1</td>\n",
       "      <td>245200915</td>\n",
       "      <td>245200929</td>\n",
       "      <td>-</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_ATCAACAGGGGTCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480301</th>\n",
       "      <td>chr1</td>\n",
       "      <td>245560998</td>\n",
       "      <td>245561012</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_AACAGAAGGTGGCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480302</th>\n",
       "      <td>chr1</td>\n",
       "      <td>246738775</td>\n",
       "      <td>246738789</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_CCCGCGAGGTGGAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480303</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247423982</td>\n",
       "      <td>247423996</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_AGCAGGAGAGGGAGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480304</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247470114</td>\n",
       "      <td>247470128</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_ATCACAAGAGGTCAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>480305 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome      Start        End strand    score  \\\n",
       "0           chr10      11458      11472      +  24.2951   \n",
       "1           chr12     547112     547126      -  24.2951   \n",
       "2            chr1     869920     869934      +  24.2951   \n",
       "3            chr1     904770     904784      +  24.2951   \n",
       "4            chr6    1397165    1397179      -  24.2951   \n",
       "...           ...        ...        ...    ...      ...   \n",
       "480300       chr1  245200915  245200929      -  10.0984   \n",
       "480301       chr1  245560998  245561012      +  10.0984   \n",
       "480302       chr1  246738775  246738789      +  10.0984   \n",
       "480303       chr1  247423982  247423996      +  10.0984   \n",
       "480304       chr1  247470114  247470128      +  10.0984   \n",
       "\n",
       "                                       name  \n",
       "0       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "1       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "2       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "3       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "4       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "...                                     ...  \n",
       "480300  CTCF_4.49e-05_0.548_ATCAACAGGGGTCAC  \n",
       "480301  CTCF_4.49e-05_0.548_AACAGAAGGTGGCCC  \n",
       "480302  CTCF_4.49e-05_0.548_CCCGCGAGGTGGAGG  \n",
       "480303  CTCF_4.49e-05_0.548_AGCAGGAGAGGGAGC  \n",
       "480304  CTCF_4.49e-05_0.548_ATCACAAGAGGTCAG  \n",
       "\n",
       "[480305 rows x 6 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Concatenate columns into a new 'name' column\n",
    "df['name'] = df['motif_alt_id'].astype(str) + '_' + df['p-value'].astype(str) + '_' + df['q-value'].astype(str) + '_' + df['matched_sequence']\n",
    "\n",
    "# Drop the original columns\n",
    "df.drop(['motif_id', 'p-value', 'q-value', 'matched_sequence', 'motif_alt_id' ], axis=1, inplace=True)\n",
    "\n",
    "# Display the updated DataFrame with the new 'name' column and without the original col\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "353295c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Chromosome', 'Start', 'End', 'strand', 'score', 'name'], dtype='object')"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns=df.columns\n",
    "columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4aeb3d8",
   "metadata": {},
   "source": [
    " ## Filter from variant chromosome names\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "21ba4768",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr10\n",
      "chr12\n",
      "chr1\n",
      "chr6\n",
      "chr4\n",
      "chrX\n",
      "chr16\n",
      "chr20\n",
      "chr22\n",
      "chr19\n",
      "chr3\n",
      "chr7\n",
      "chr11\n",
      "chr8\n",
      "chr14\n",
      "chr17\n",
      "chr13\n",
      "chr2\n",
      "chr9\n",
      "hm\n",
      "chr18\n",
      "chr14_KI270726v1_random\n",
      "chr15\n",
      "chr16_KI270728v1_random\n",
      "chr21\n",
      "chr22_KI270734v1_random\n",
      "chr22_KI270731v1_random\n",
      "chr1_KI270714v1_random\n",
      "chr15_KI270727v1_random\n",
      "chrEBV\n",
      "chr11_KI270721v1_random\n",
      "chr14_KI270723v1_random\n",
      "chr17_GL000205v2_random\n",
      "chrUn_GL000218v1\n",
      "chr14_GL000194v1_random\n",
      "chrUn_KI270744v1\n",
      "chrUn_GL000213v1\n",
      "chrUn_KI270745v1\n",
      "chrUn_GL000195v1\n",
      "chrUn_GL000219v1\n",
      "chr1_KI270712v1_random\n",
      "chrUn_KI270750v1\n",
      "chrUn_KI270755v1\n",
      "chrUn_KI270753v1\n",
      "chr22_KI270733v1_random\n",
      "chr1_KI270706v1_random\n",
      "chrUn_KI270748v1\n",
      "chrUn_KI270741v1\n",
      "chr1_KI270711v1_random\n",
      "chrUn_GL000224v1\n",
      "chr1_KI270708v1_random\n",
      "chr14_KI270722v1_random\n",
      "chr1_KI270713v1_random\n",
      "chr14_KI270725v1_random\n",
      "chrUn_KI270749v1\n",
      "chr9_KI270717v1_random\n",
      "chr14_KI270724v1_random\n",
      "chrUn_KI270752v1\n",
      "chrUn_KI270435v1\n",
      "chrUn_KI270743v1\n",
      "chr4_GL000008v2_random\n",
      "chr14_GL000009v2_random\n",
      "chr5_GL000208v1_random\n",
      "chr14_GL000225v1_random\n",
      "chrUn_GL000216v2\n",
      "chrUn_GL000220v1\n",
      "chr9_KI270719v1_random\n",
      "chr9_KI270718v1_random\n",
      "chr3_GL000221v1_random\n",
      "chrUn_KI270742v1\n",
      "chrUn_KI270746v1\n",
      "chr22_KI270732v1_random\n",
      "chr1_KI270707v1_random\n",
      "chrUn_KI270747v1\n",
      "chrUn_KI270315v1\n",
      "chrUn_KI270304v1\n",
      "chrUn_GL000214v1\n",
      "chr22_KI270739v1_random\n",
      "chrUn_KI270519v1\n",
      "chrUn_KI270442v1\n",
      "chrUn_KI270751v1\n",
      "chr1_KI270709v1_random\n",
      "chrUn_KI270589v1\n",
      "chrUn_KI270757v1\n",
      "chrUn_KI270371v1\n",
      "chrUn_KI270511v1\n",
      "chr1_KI270710v1_random\n",
      "chrUn_KI270754v1\n",
      "chr17_KI270729v1_random\n",
      "chrUn_KI270539v1\n",
      "chr22_KI270735v1_random\n",
      "chr22_KI270738v1_random\n",
      "chr22_KI270736v1_random\n",
      "chr9_KI270720v1_random\n",
      "chr22_KI270737v1_random\n",
      "chrUn_KI270530v1\n",
      "chrY_KI270740v1_random\n",
      "chrUn_KI270538v1\n",
      "chrUn_KI270419v1\n",
      "chrUn_KI270423v1\n",
      "chrUn_KI270422v1\n",
      "chrUn_KI270418v1\n",
      "chr17_KI270730v1_random\n"
     ]
    }
   ],
   "source": [
    "#Discover the unique sequence names in fimo.tsv\n",
    "for sequence_name in df['Chromosome'].unique():\n",
    "    print(sequence_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "352ec955",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Expected list of NCBI values\n",
    "expected_list = ['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', 'chr11', 'chr12',\n",
    "                 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chrX', 'chrY', 'chrM']\n",
    "\n",
    "# Filter rows based on 'NCBI' column values matching the expected list\n",
    "df = df[df['Chromosome'].isin(expected_list)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "b61300cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>11458</td>\n",
       "      <td>11472</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>547112</td>\n",
       "      <td>547126</td>\n",
       "      <td>-</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>869920</td>\n",
       "      <td>869934</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>904770</td>\n",
       "      <td>904784</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>1397165</td>\n",
       "      <td>1397179</td>\n",
       "      <td>-</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480300</th>\n",
       "      <td>chr1</td>\n",
       "      <td>245200915</td>\n",
       "      <td>245200929</td>\n",
       "      <td>-</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_ATCAACAGGGGTCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480301</th>\n",
       "      <td>chr1</td>\n",
       "      <td>245560998</td>\n",
       "      <td>245561012</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_AACAGAAGGTGGCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480302</th>\n",
       "      <td>chr1</td>\n",
       "      <td>246738775</td>\n",
       "      <td>246738789</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_CCCGCGAGGTGGAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480303</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247423982</td>\n",
       "      <td>247423996</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_AGCAGGAGAGGGAGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480304</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247470114</td>\n",
       "      <td>247470128</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_ATCACAAGAGGTCAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>451732 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome      Start        End strand    score  \\\n",
       "0           chr10      11458      11472      +  24.2951   \n",
       "1           chr12     547112     547126      -  24.2951   \n",
       "2            chr1     869920     869934      +  24.2951   \n",
       "3            chr1     904770     904784      +  24.2951   \n",
       "4            chr6    1397165    1397179      -  24.2951   \n",
       "...           ...        ...        ...    ...      ...   \n",
       "480300       chr1  245200915  245200929      -  10.0984   \n",
       "480301       chr1  245560998  245561012      +  10.0984   \n",
       "480302       chr1  246738775  246738789      +  10.0984   \n",
       "480303       chr1  247423982  247423996      +  10.0984   \n",
       "480304       chr1  247470114  247470128      +  10.0984   \n",
       "\n",
       "                                       name  \n",
       "0       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "1       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "2       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "3       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "4       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "...                                     ...  \n",
       "480300  CTCF_4.49e-05_0.548_ATCAACAGGGGTCAC  \n",
       "480301  CTCF_4.49e-05_0.548_AACAGAAGGTGGCCC  \n",
       "480302  CTCF_4.49e-05_0.548_CCCGCGAGGTGGAGG  \n",
       "480303  CTCF_4.49e-05_0.548_AGCAGGAGAGGGAGC  \n",
       "480304  CTCF_4.49e-05_0.548_ATCACAAGAGGTCAG  \n",
       "\n",
       "[451732 rows x 6 columns]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8f6a3a5",
   "metadata": {},
   "source": [
    "## Save df as .bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "987d5305",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the csv file \n",
    "df.to_csv('fimo_full.bed', sep='\\t', index=False, header=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c57b6c4",
   "metadata": {},
   "source": [
    "# Mart to bed\n",
    "Convert mart_export.txt to bed format "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "665a1a65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Start (bp)</th>\n",
       "      <th>End (bp)</th>\n",
       "      <th>Feature type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18</td>\n",
       "      <td>35116801</td>\n",
       "      <td>35120999</td>\n",
       "      <td>Enhancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6</td>\n",
       "      <td>90249202</td>\n",
       "      <td>90257999</td>\n",
       "      <td>Enhancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>39894801</td>\n",
       "      <td>39895400</td>\n",
       "      <td>Enhancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15</td>\n",
       "      <td>63059401</td>\n",
       "      <td>63062399</td>\n",
       "      <td>Enhancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>119868001</td>\n",
       "      <td>119868400</td>\n",
       "      <td>Enhancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305075</th>\n",
       "      <td>8</td>\n",
       "      <td>98939200</td>\n",
       "      <td>98949401</td>\n",
       "      <td>Promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305076</th>\n",
       "      <td>4</td>\n",
       "      <td>39364800</td>\n",
       "      <td>39367201</td>\n",
       "      <td>Promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305077</th>\n",
       "      <td>X</td>\n",
       "      <td>152752601</td>\n",
       "      <td>152755400</td>\n",
       "      <td>Promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305078</th>\n",
       "      <td>16</td>\n",
       "      <td>58495400</td>\n",
       "      <td>58496001</td>\n",
       "      <td>Promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305079</th>\n",
       "      <td>7</td>\n",
       "      <td>38332400</td>\n",
       "      <td>38337001</td>\n",
       "      <td>Promoter</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>305080 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome/scaffold name  Start (bp)   End (bp) Feature type\n",
       "0                            18    35116801   35120999     Enhancer\n",
       "1                             6    90249202   90257999     Enhancer\n",
       "2                             7    39894801   39895400     Enhancer\n",
       "3                            15    63059401   63062399     Enhancer\n",
       "4                             3   119868001  119868400     Enhancer\n",
       "...                         ...         ...        ...          ...\n",
       "305075                        8    98939200   98949401     Promoter\n",
       "305076                        4    39364800   39367201     Promoter\n",
       "305077                        X   152752601  152755400     Promoter\n",
       "305078                       16    58495400   58496001     Promoter\n",
       "305079                        7    38332400   38337001     Promoter\n",
       "\n",
       "[305080 rows x 4 columns]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the TSV file into a DataFrame\n",
    "#file_path=input(\"Give mart_export_unique.txt file path: \")\n",
    "file_path = '/mnt/raid1/thalassini/home/Downloads/mart_export.txt'\n",
    "mart_df = pd.read_csv(file_path, sep='\\t')\n",
    "mart_df.dropna(inplace=True)\n",
    "\n",
    "# Display the DataFrame\n",
    "mart_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "6cd07e6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapping table\n",
    "mapping_table = {\n",
    "    '1': 'chr1',\n",
    "    '2': 'chr2',\n",
    "    '3': 'chr3',\n",
    "    '4': 'chr4',\n",
    "    '5': 'chr5',\n",
    "    '6': 'chr6',\n",
    "    '7': 'chr7',\n",
    "    '8': 'chr8',\n",
    "    '9': 'chr9',\n",
    "    '10': 'chr10',\n",
    "    '11': 'chr11',\n",
    "    '12': 'chr12',\n",
    "    '13': 'chr13',\n",
    "    '14': 'chr14',\n",
    "    '15': 'chr15',\n",
    "    '16': 'chr16',\n",
    "    '17': 'chr17',\n",
    "    '18': 'chr18',\n",
    "    '19': 'chr19',\n",
    "    '20': 'chr20',\n",
    "    '21': 'chr21',\n",
    "    '22': 'chr22',\n",
    "    'X': 'chrX',\n",
    "    'Y': 'chrY',\n",
    "    'MT': 'chrM'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "f3dcf4b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Start (bp)</th>\n",
       "      <th>End (bp)</th>\n",
       "      <th>Feature type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr18</td>\n",
       "      <td>35116801</td>\n",
       "      <td>35120999</td>\n",
       "      <td>Enhancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>90249202</td>\n",
       "      <td>90257999</td>\n",
       "      <td>Enhancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7</td>\n",
       "      <td>39894801</td>\n",
       "      <td>39895400</td>\n",
       "      <td>Enhancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr15</td>\n",
       "      <td>63059401</td>\n",
       "      <td>63062399</td>\n",
       "      <td>Enhancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3</td>\n",
       "      <td>119868001</td>\n",
       "      <td>119868400</td>\n",
       "      <td>Enhancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305075</th>\n",
       "      <td>chr8</td>\n",
       "      <td>98939200</td>\n",
       "      <td>98949401</td>\n",
       "      <td>Promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305076</th>\n",
       "      <td>chr4</td>\n",
       "      <td>39364800</td>\n",
       "      <td>39367201</td>\n",
       "      <td>Promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305077</th>\n",
       "      <td>chrX</td>\n",
       "      <td>152752601</td>\n",
       "      <td>152755400</td>\n",
       "      <td>Promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305078</th>\n",
       "      <td>chr16</td>\n",
       "      <td>58495400</td>\n",
       "      <td>58496001</td>\n",
       "      <td>Promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305079</th>\n",
       "      <td>chr7</td>\n",
       "      <td>38332400</td>\n",
       "      <td>38337001</td>\n",
       "      <td>Promoter</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>305080 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome/scaffold name  Start (bp)   End (bp) Feature type\n",
       "0                         chr18    35116801   35120999     Enhancer\n",
       "1                          chr6    90249202   90257999     Enhancer\n",
       "2                          chr7    39894801   39895400     Enhancer\n",
       "3                         chr15    63059401   63062399     Enhancer\n",
       "4                          chr3   119868001  119868400     Enhancer\n",
       "...                         ...         ...        ...          ...\n",
       "305075                     chr8    98939200   98949401     Promoter\n",
       "305076                     chr4    39364800   39367201     Promoter\n",
       "305077                     chrX   152752601  152755400     Promoter\n",
       "305078                    chr16    58495400   58496001     Promoter\n",
       "305079                     chr7    38332400   38337001     Promoter\n",
       "\n",
       "[305080 rows x 4 columns]"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mart_df['Chromosome/scaffold name'] = mart_df['Chromosome/scaffold name'].map(mapping_table)\n",
    "mart_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "66653438",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the csv file \n",
    "mart_df.to_csv('mart.bed', sep='\\t', index=False, header=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99080618",
   "metadata": {},
   "source": [
    "# Run bedtools intesect to create filtered fimo\n",
    ">> bedtools intersect -a /mnt/raid1/thalassini/home/Downloads/fimo_full.bed  -b /mnt/raid1/thalassini/home/Downloads/mart.bed > ensemlb_filtered_fimo.bed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27daf313",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f88f56bd",
   "metadata": {},
   "source": [
    "# Display filtered_fimo df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "b06eaa90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>869920</td>\n",
       "      <td>869934</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>904770</td>\n",
       "      <td>904784</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6</td>\n",
       "      <td>1397165</td>\n",
       "      <td>1397179</td>\n",
       "      <td>-</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr4</td>\n",
       "      <td>1579280</td>\n",
       "      <td>1579294</td>\n",
       "      <td>-</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>9995929</td>\n",
       "      <td>9995943</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104638</th>\n",
       "      <td>chr2</td>\n",
       "      <td>239339669</td>\n",
       "      <td>239339683</td>\n",
       "      <td>-</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_GCCAGGAGATTGCAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104639</th>\n",
       "      <td>chr2</td>\n",
       "      <td>241355101</td>\n",
       "      <td>241355115</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_GCCACATGGTGACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104640</th>\n",
       "      <td>chr1</td>\n",
       "      <td>241396943</td>\n",
       "      <td>241396957</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_GCCACATGGTGACAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104641</th>\n",
       "      <td>chr2</td>\n",
       "      <td>241764826</td>\n",
       "      <td>241764840</td>\n",
       "      <td>-</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_AGCAGGAGAGGGAGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104642</th>\n",
       "      <td>chr1</td>\n",
       "      <td>245200915</td>\n",
       "      <td>245200929</td>\n",
       "      <td>-</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_ATCAACAGGGGTCAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104643 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome      Start        End strand    score  \\\n",
       "0            chr1     869920     869934      +  24.2951   \n",
       "1            chr1     904770     904784      +  24.2951   \n",
       "2            chr6    1397165    1397179      -  24.2951   \n",
       "3            chr4    1579280    1579294      -  24.2951   \n",
       "4            chrX    9995929    9995943      +  24.2951   \n",
       "...           ...        ...        ...    ...      ...   \n",
       "104638       chr2  239339669  239339683      -  10.0984   \n",
       "104639       chr2  241355101  241355115      +  10.0984   \n",
       "104640       chr1  241396943  241396957      +  10.0984   \n",
       "104641       chr2  241764826  241764840      -  10.0984   \n",
       "104642       chr1  245200915  245200929      -  10.0984   \n",
       "\n",
       "                                       name  \n",
       "0       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "1       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "2       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "3       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "4       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "...                                     ...  \n",
       "104638  CTCF_4.49e-05_0.548_GCCAGGAGATTGCAG  \n",
       "104639  CTCF_4.49e-05_0.548_GCCACATGGTGACAG  \n",
       "104640  CTCF_4.49e-05_0.548_GCCACATGGTGACAG  \n",
       "104641  CTCF_4.49e-05_0.548_AGCAGGAGAGGGAGC  \n",
       "104642  CTCF_4.49e-05_0.548_ATCAACAGGGGTCAC  \n",
       "\n",
       "[104643 rows x 6 columns]"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the BED file into a DataFrame\n",
    "#file_path=input(\"Give ensemlb_filtered_fimo.bed file path: \")\n",
    "file_path = '/mnt/raid1/thalassini/home/ensemlb_filtered_fimo.bed'\n",
    "filtered_df = pd.read_csv(file_path, sep='\\t')\n",
    "filtered_df.dropna(inplace=True)\n",
    "filtered_df.columns=columns\n",
    "# Display the DataFrame\n",
    "filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "ecf5d6f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " The promoter/enhancer filtered data is reduced by 76.8351588995245% from initial fimo results\n"
     ]
    }
   ],
   "source": [
    "print(f\" The promoter/enhancer filtered data is reduced by {100-len(filtered_df)* 100/len(df)}% from initial fimo results\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05a2044f",
   "metadata": {},
   "source": [
    "# Run bedtools intesect to create chipseq_filtered fimo\n",
    ">> bedtools intersect -a ensemlb_filtered_fimo.bed -b /mnt/raid1/thalassini/home/Downloads/ENCFF970MXD.bed > chipseq_filtered.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cb7f978",
   "metadata": {},
   "source": [
    "# Display chipseq_filtered df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "dbea56ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>869920</td>\n",
       "      <td>869934</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>904770</td>\n",
       "      <td>904784</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6</td>\n",
       "      <td>1397165</td>\n",
       "      <td>1397179</td>\n",
       "      <td>-</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr4</td>\n",
       "      <td>1579280</td>\n",
       "      <td>1579294</td>\n",
       "      <td>-</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>9995929</td>\n",
       "      <td>9995943</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17931</th>\n",
       "      <td>chr4</td>\n",
       "      <td>112637587</td>\n",
       "      <td>112637601</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_GCCAGAAGATAGCGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17932</th>\n",
       "      <td>chr9</td>\n",
       "      <td>122810697</td>\n",
       "      <td>122810711</td>\n",
       "      <td>-</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_ACCACAATAGGGCAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17933</th>\n",
       "      <td>chr9</td>\n",
       "      <td>122810697</td>\n",
       "      <td>122810711</td>\n",
       "      <td>-</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_ACCACAATAGGGCAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17934</th>\n",
       "      <td>chr7</td>\n",
       "      <td>131581578</td>\n",
       "      <td>131581592</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_GCCAACAAAGGGCAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17935</th>\n",
       "      <td>chr8</td>\n",
       "      <td>143771853</td>\n",
       "      <td>143771867</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_AGCAGGAGGAGGCGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17936 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome      Start        End strand    score  \\\n",
       "0           chr1     869920     869934      +  24.2951   \n",
       "1           chr1     904770     904784      +  24.2951   \n",
       "2           chr6    1397165    1397179      -  24.2951   \n",
       "3           chr4    1579280    1579294      -  24.2951   \n",
       "4           chrX    9995929    9995943      +  24.2951   \n",
       "...          ...        ...        ...    ...      ...   \n",
       "17931       chr4  112637587  112637601      +  10.0984   \n",
       "17932       chr9  122810697  122810711      -  10.0984   \n",
       "17933       chr9  122810697  122810711      -  10.0984   \n",
       "17934       chr7  131581578  131581592      +  10.0984   \n",
       "17935       chr8  143771853  143771867      +  10.0984   \n",
       "\n",
       "                                      name  \n",
       "0      CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "1      CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "2      CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "3      CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "4      CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "...                                    ...  \n",
       "17931  CTCF_4.49e-05_0.548_GCCAGAAGATAGCGA  \n",
       "17932  CTCF_4.49e-05_0.548_ACCACAATAGGGCAG  \n",
       "17933  CTCF_4.49e-05_0.548_ACCACAATAGGGCAG  \n",
       "17934  CTCF_4.49e-05_0.548_GCCAACAAAGGGCAG  \n",
       "17935  CTCF_4.49e-05_0.548_AGCAGGAGGAGGCGG  \n",
       "\n",
       "[17936 rows x 6 columns]"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the BED file into a DataFrame\n",
    "#file_path=input(\"Give chipseq_filtered_fimo.bed file path: \")\n",
    "file_path = '/mnt/raid1/thalassini/home/chipseq_filtered.bed'\n",
    "chipseq_filtered_df = pd.read_csv(file_path, sep='\\t')\n",
    "chipseq_filtered_df.dropna(inplace=True)\n",
    "chipseq_filtered_df.columns=columns\n",
    "# Display the DataFrame\n",
    "chipseq_filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "e7d3f011",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        chr1\n",
       "1        chr1\n",
       "2        chr6\n",
       "3        chr4\n",
       "4        chrX\n",
       "         ... \n",
       "17931    chr4\n",
       "17932    chr9\n",
       "17933    chr9\n",
       "17934    chr7\n",
       "17935    chr8\n",
       "Name: Chromosome, Length: 17936, dtype: object"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for chromosome in chipseq_filtered_df[\"Chromosome\"]:\n",
    "    print(chipseq_filtered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "47c6a29e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " The chipseq filtered data is reduced by 96.02950421931588% from initial fimo results\n"
     ]
    }
   ],
   "source": [
    "print(f\" The chipseq filtered data is reduced by {100-len(chipseq_filtered_df)* 100/len(df)}% from initial fimo results\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9ff85ec",
   "metadata": {},
   "source": [
    "# Find sequence from coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "fb3c0a26",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "odict_keys(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', 'X', 'Y', 'MT', 'HG76_PATCH', 'HG2365_PATCH', 'HSCHR15_4_CTG8', 'HSCHR6_MHC_SSTO_CTG1', 'HSCHR6_MHC_MCF_CTG1', 'HSCHR6_MHC_COX_CTG1', 'HSCHR6_MHC_MANN_CTG1', 'HSCHR6_MHC_APD_CTG1', 'HSCHR6_MHC_QBL_CTG1', 'HSCHR6_MHC_DBB_CTG1', 'HSCHR17_7_CTG4', 'HSCHR16_1_CTG1', 'HSCHR15_6_CTG8', 'HG926_PATCH', 'HSCHR17_1_CTG5', 'HSCHR5_2_CTG1_1', 'HG1343_HG173_HG459_PATCH', 'HSCHR14_7_CTG1', 'HSCHR17_2_CTG5', 'HSCHR14_3_CTG1', 'HG2280_PATCH', 'HSCHR5_1_CTG1_1', 'HSCHR7_2_CTG6', 'HSCHR19LRC_LRC_J_CTG3_1', 'HSCHR19_4_CTG3_1', 'HSCHR19LRC_LRC_S_CTG3_1', 'HSCHR19LRC_LRC_I_CTG3_1', 'HG1815_PATCH', 'HG2265_PATCH', 'HSCHR19LRC_LRC_T_CTG3_1', 'HSCHR19LRC_COX1_CTG3_1', 'HSCHR19LRC_PGF1_CTG3_1', 'HSCHR8_8_CTG1', 'HG2275_PATCH', 'HSCHR1_3_CTG32_1', 'HSCHR6_1_CTG8', 'HG1532_PATCH', 'HSCHR1_5_CTG31', 'HSCHR19LRC_PGF2_CTG3_1', 'HSCHR19LRC_COX2_CTG3_1', 'HG2526_HG2573_PATCH', 'HG1_PATCH', 'HSCHR7_3_CTG4_4', 'HG30_PATCH', 'HSCHR8_9_CTG1', 'HSCHRX_1_CTG14', 'HG708_PATCH', 'HSCHR4_1_CTG9', 'HSCHR12_3_CTG2', 'HG1206_PATCH', 'HSCHR4_7_CTG12', 'HSCHR1_6_CTG3', 'HG1362_PATCH', 'HG2407_PATCH', 'HSCHR2_11_CTG7_2', 'HSCHR2_6_CTG7_2', 'HG2513_PATCH', 'HSCHR1_12_CTG3', 'HG2308_PATCH', 'HSCHR2_12_CTG7_2', 'HG109_PATCH', 'HG2052_PATCH', 'HG1012_PATCH', 'HG2290_PATCH', 'HG2263_PATCH', 'HSCHR15_1_CTG1', 'HG2069_PATCH', 'HG2264_PATCH', 'HG2266_PATCH', 'HG1342_HG2282_PATCH', 'HSCHR7_4_CTG1', 'HG1485_PATCH', 'HG1832_PATCH', 'HG545_PATCH', 'HG2066_PATCH', 'HSCHR9_1_CTG5', 'HG152_PATCH', 'HSCHR2_10_CTG7_2', 'HSCHR15_5_CTG8', 'HG2267_PATCH', 'HSCHR1_1_CTG32_1', 'HSCHR4_12_CTG12', 'HG2244_HG2245_PATCH', 'HG126_PATCH', 'HG2512_PATCH', 'HG2236_PATCH', 'HG2509_PATCH', 'HSCHR12_2_CTG2', 'HG2213_PATCH', 'HG26_PATCH', 'HG439_PATCH', 'HG2510_PATCH', 'HG2511_PATCH', 'HSCHR17_2_CTG2', 'HSCHR15_2_CTG8', 'HSCHR19_1_CTG2', 'HSCHR6_1_CTG1', 'HSCHR4_6_CTG12', 'HSCHR4_1_CTG6', 'HSCHR17_1_CTG1', 'HSCHR8_1_CTG7', 'HG2235_PATCH', 'HSCHR17_13_CTG4', 'HG2021_PATCH', 'HG2198_PATCH', 'HSCHR1_3_CTG31', 'HG2232_PATCH', 'HSCHR5_8_CTG1', 'HG2291_PATCH', 'HG410_PATCH', 'HSCHR1_1_CTG3', 'HSCHR1_4_CTG3', 'HG2568_PATCH', 'HG2525_PATCH', 'HG2158_PATCH', 'HG2219_PATCH', 'HSCHR1_6_CTG31', 'HG405_PATCH', 'HG2461_PATCH', 'HSCHRX_3_CTG3', 'HG2030_PATCH', 'HG2002_PATCH', 'HSCHR15_1_CTG3', 'HSCHR17_10_CTG4', 'HSCHR14_1_CTG1', 'HG1651_PATCH', 'HSCHR8_5_CTG1', 'HG2063_PATCH', 'HSCHR10_1_CTG2', 'HSCHR4_2_CTG8_1', 'HSCHR13_1_CTG1', 'HG2022_PATCH', 'HSCHR8_3_CTG7', 'HG2095_PATCH', 'HG2060_PATCH', 'HSCHR22_1_CTG7', 'HSCHR3_9_CTG2_1', 'HSCHR5_10_CTG1', 'HSCHR11_2_CTG8', 'HG2077_PATCH', 'HSCHR8_3_CTG1', 'HG2246_HG2248_HG2276_PATCH', 'HSCHR11_1_CTG7', 'HSCHR15_1_CTG8', 'HSCHRX_2_CTG14', 'HSCHR19KIR_CA01-TB04_CTG3_1', 'HG1277_PATCH', 'HSCHR8_1_CTG1', 'HSCHR18_1_CTG1_1', 'HSCHR2_6_CTG1', 'HSCHRX_1_CTG3', 'HSCHR8_2_CTG1', 'HSCHR19KIR_FH05_B_HAP_CTG3_1', 'HG2285_HG106_HG2252_PATCH', 'HG2114_PATCH', 'HSCHR1_5_CTG3', 'HSCHR17_4_CTG4', 'HG2191_PATCH', 'HSCHR17_3_CTG1', 'HG705_PATCH', 'HSCHRX_2_CTG3', 'HG2225_PATCH', 'HSCHR6_1_CTG5', 'HSCHR7_3_CTG6', 'HG2471_PATCH', 'HG28_PATCH', 'HG2128_PATCH', 'HSCHR16_4_CTG3_1', 'HG2068_PATCH', 'HG2334_PATCH', 'HSCHR14_9_CTG1', 'HSCHR22_1_CTG4', 'HSCHR15_3_CTG3', 'HSCHR11_2_CTG3_1', 'HSCHR22_1_CTG3', 'HSCHR1_2_CTG3', 'HG1309_PATCH', 'HG2577_PATCH', 'HSCHR19KIR_FH15_B_HAP_CTG3_1', 'HSCHR3_3_CTG2_1', 'HSCHR17_11_CTG4', 'HG2072_PATCH', 'HSCHR15_3_CTG8', 'HG2231_HG2496_PATCH', 'HSCHR4_3_CTG12', 'HG2057_PATCH', 'HSCHR19KIR_CA01-TB01_CTG3_1', 'HSCHR12_4_CTG2', 'HSCHR4_9_CTG12', 'HSCHR17_8_CTG4', 'HG699_PATCH', 'HG2118_PATCH', 'HSCHR19_3_CTG3_1', 'HSCHR16_3_CTG1', 'HG2412_PATCH', 'HG2023_PATCH', 'HSCHR8_7_CTG1', 'HSCHR5_2_CTG1', 'HSCHR3_4_CTG2_1', 'HSCHR17_2_CTG4', 'HSCHR2_4_CTG1', 'HSCHR19KIR_0019-4656-B_CTG3_1', 'HSCHR4_1_CTG4', 'HSCHR11_3_CTG1', 'HG2115_PATCH', 'HSCHR19KIR_FH13_BA2_HAP_CTG3_1', 'HSCHR3_5_CTG1', 'HSCHR11_2_CTG1_1', 'HSCHR2_2_CTG7', 'HSCHR1_5_CTG32_1', 'HG2515_PATCH', 'HG23_PATCH', 'HG2571_PATCH', 'HSCHR11_1_CTG8', 'HSCHR7_1_CTG7', 'HG1535_PATCH', 'HSCHR7_2_CTG4_4', 'HSCHR19KIR_GRC212_AB_HAP_CTG3_1', 'HG2476_PATCH', 'HG2233_PATCH', 'HG2031_PATCH', 'HG2288_HG2289_PATCH', 'HSCHR4_5_CTG12', 'HG287_PATCH', 'HSCHR4_11_CTG12', 'HSCHR3_1_CTG3', 'HSCHR19KIR_RSH_BA2_HAP_CTG3_1', 'HSCHR18_5_CTG1_1', 'HG107_HG2565_PATCH', 'HSCHR19KIR_G248_BA2_HAP_CTG3_1', 'HSCHR11_1_CTG5', 'HSCHR15_9_CTG8', 'HSCHR11_1_CTG1_2', 'HSCHR21_2_CTG1_1', 'HG142_HG150_NOVEL_TEST', 'HSCHR19KIR_FH08_BAX_HAP_CTG3_1', 'HSCHR18_2_CTG2', 'HSCHR19KIR_FH05_A_HAP_CTG3_1', 'HSCHR3_8_CTG2_1', 'HSCHR6_1_CTG4', 'HSCHR3_6_CTG3', 'HSCHR17_1_CTG2', 'HSCHR15_2_CTG3', 'HG2576_PATCH', 'HSCHR5_3_CTG1', 'HSCHR3_6_CTG2_1', 'HG1398_PATCH', 'HSCHR16_1_CTG3_1', 'HSCHR13_1_CTG3', 'HG151_NOVEL_TEST', 'HSCHR19KIR_ABC08_A1_HAP_CTG3_1', 'HSCHR7_2_CTG1', 'HG2521_PATCH', 'HSCHR19KIR_FH06_BA1_HAP_CTG3_1', 'HSCHR10_1_CTG4', 'HSCHR19_3_CTG2', 'HSCHRX_3_CTG7', 'HSCHR19KIR_FH06_A_HAP_CTG3_1', 'HG1369_PATCH', 'HG986_PATCH', 'HSCHR22_1_CTG6', 'HSCHR19KIR_FH08_A_HAP_CTG3_1', 'HSCHR11_1_CTG2', 'HSCHR6_1_CTG3', 'HSCHR10_1_CTG6', 'HSCHR1_4_CTG31', 'HSCHR19KIR_T7526_BDEL_HAP_CTG3_1', 'HSCHR19KIR_LUCE_BDEL_HAP_CTG3_1', 'HSCHR3_9_CTG3', 'HSCHR12_4_CTG2_1', 'HSCHR20_1_CTG3', 'HSCHR1_1_CTG31', 'HG2116_PATCH', 'HSCHR2_7_CTG7_2', 'HSCHR10_1_CTG3', 'HSCHR11_1_CTG3_1', 'HSCHR14_2_CTG1', 'HSCHR3_1_CTG2_1', 'HSCHR13_1_CTG2', 'HG2239_PATCH', 'HSCHR10_1_CTG1', 'HSCHR5_4_CTG1', 'HSCHR17_1_CTG9', 'HSCHR19KIR_502960008-2_CTG3_1', 'HSCHR19KIR_RP5_B_HAP_CTG3_1', 'HSCHR11_1_CTG6', 'HG1298_PATCH', 'HSCHR22_3_CTG1', 'HSCHR3_4_CTG1', 'HSCHR6_1_CTG7', 'HSCHR12_8_CTG2_1', 'HSCHR22_7_CTG1', 'HG2140_PATCH', 'HSCHR2_3_CTG15', 'HSCHR18_ALT21_CTG2_1', 'HSCHR3_2_CTG2_1', 'HSCHR5_1_CTG5', 'HSCHR3_1_CTG1', 'HSCHR5_4_CTG1_1', 'HG2580_PATCH', 'HSCHR7_3_CTG1', 'HSCHR14_8_CTG1', 'HSCHR9_1_CTG3', 'HSCHR19KIR_7191059-2_CTG3_1', 'HSCHR19KIR_G248_A_HAP_CTG3_1', 'HG2111_PATCH', 'HSCHR19KIR_ABC08_AB_HAP_T_P_CTG3_1', 'HSCHR19KIR_LUCE_A_HAP_CTG3_1', 'HSCHR19KIR_RSH_A_HAP_CTG3_1', 'HSCHR19KIR_T7526_A_HAP_CTG3_1', 'HSCHR19KIR_FH15_A_HAP_CTG3_1', 'HSCHR19KIR_G085_A_HAP_CTG3_1', 'HSCHR19_2_CTG2', 'HSCHR19KIR_CA01-TA01_1_CTG3_1', 'HSCHR19KIR_FH13_A_HAP_CTG3_1', 'HSCHR12_1_CTG2_1', 'HSCHR16_3_CTG3_1', 'HSCHR13_1_CTG4', 'HSCHR13_1_CTG7', 'HSCHR19KIR_CA01-TA01_2_CTG3_1', 'HSCHR18_3_CTG2_1', 'HSCHR18_1_CTG2_1', 'HSCHR12_1_CTG1', 'HSCHR21_8_CTG1_1', 'HSCHR19KIR_502960008-1_CTG3_1', 'HSCHR3_4_CTG3', 'HG2058_PATCH', 'HSCHR1_8_CTG3', 'HSCHR1_2_CTG32_1', 'HG2237_PATCH', 'HSCHR3_8_CTG3', 'HG1445_PATCH', 'HG2419_PATCH', 'HSCHR18_2_CTG1_1', 'HSCHR5_6_CTG1', 'HSCHR4_1_CTG12', 'HSCHR3_7_CTG3', 'HSCHR3_7_CTG2_1', 'HSCHR19KIR_0019-4656-A_CTG3_1', 'HSCHR9_1_CTG6', 'HSCHR4_8_CTG12', 'HSCHR3_5_CTG3', 'HSCHR18_1_CTG1', 'HSCHR9_1_CTG1', 'HSCHR3_3_CTG3', 'HSCHR22_1_CTG1', 'HSCHR3_2_CTG3', 'HSCHR1_ALT2_1_CTG32_1', 'HSCHR2_1_CTG15', 'HSCHR2_2_CTG15', 'HSCHR19KIR_7191059-1_CTG3_1', 'HSCHR18_2_CTG2_1', 'HSCHR19KIR_HG2394_CTG3_1', 'HSCHR8_4_CTG7', 'HSCHR4_4_CTG12', 'HG1523_PATCH', 'HSCHR7_1_CTG4_4', 'HSCHR7_1_CTG1', 'HSCHR18_ALT2_CTG2_1', 'HSCHR17_3_CTG2', 'HSCHR19KIR_G085_BA1_HAP_CTG3_1', 'HG2067_PATCH', 'HSCHR19KIR_HG2393_CTG3_1', 'HSCHR22_4_CTG1', 'HSCHR22_6_CTG1', 'HSCHR19_1_CTG3_1', 'HSCHR19KIR_GRC212_BA1_HAP_CTG3_1', 'HG2046_PATCH', 'HSCHR11_1_CTG1_1', 'HSCHR12_9_CTG2_1', 'HSCHR22_5_CTG1', 'HSCHR12_3_CTG2_1', 'HSCHR6_1_CTG6', 'HG2176_PATCH', 'HG2249_PATCH', 'HG721_PATCH', 'HSCHR19KIR_HG2396_CTG3_1', 'HSCHR8_4_CTG1', 'HSCHR22_8_CTG1', 'HG2251_PATCH', 'HSCHRX_2_CTG12', 'HSCHR21_6_CTG1_1', 'HSCHR2_1_CTG5', 'HG2088_PATCH', 'HSCHR8_1_CTG6', 'HSCHR1_3_CTG3', 'HG1708_PATCH', 'HG2541_PATCH', 'HSCHR2_8_CTG7_2', 'HSCHR1_9_CTG3', 'HG2121_PATCH', 'HSCHR5_7_CTG1', 'HSCHR12_2_CTG2_1', 'HSCHR2_5_CTG7_2', 'HG2087_PATCH', 'HG2569_PATCH', 'HSCHR17_2_CTG1', 'HSCHR8_6_CTG1', 'HSCHR2_2_CTG1', 'HSCHR5_9_CTG1', 'HSCHR16_CTG2', 'HSCHR9_1_CTG7', 'HSCHR8_5_CTG7', 'HSCHR17_1_CTG4', 'HSCHR2_1_CTG7', 'HSCHR8_2_CTG7', 'HSCHR5_2_CTG5', 'HSCHR5_3_CTG5', 'HSCHR20_1_CTG1', 'HSCHR7_2_CTG7', 'HSCHR5_5_CTG1', 'HG1296_PATCH', 'HSCHR6_1_CTG2', 'HSCHR2_1_CTG7_2', 'HSCHR13_1_CTG8', 'HSCHR19KIR_ABC08_AB_HAP_C_P_CTG3_1', 'HG2578_PATCH', 'HG2104_PATCH', 'HSCHR12_1_CTG2', 'HSCHR2_1_CTG1', 'HG2494_PATCH', 'HSCHR4_2_CTG12', 'HSCHR12_6_CTG2_1', 'HSCHR7_1_CTG6', 'HSCHR20_1_CTG2', 'HSCHR17_12_CTG4', 'HSCHR21_4_CTG1_1', 'HG2155_PATCH', 'HSCHR3_5_CTG2_1', 'HSCHR4_1_CTG8_1', 'HSCHR18_4_CTG1_1', 'HSCHR2_4_CTG7_2', 'HSCHR1_2_CTG31', 'HSCHR2_3_CTG7_2', 'HSCHR1_1_CTG11', 'HSCHR12_2_CTG1', 'HSCHR3_3_CTG1', 'HG2217_PATCH', 'HSCHR17_9_CTG4', 'HG2408_PATCH', 'HSCHR11_2_CTG1', 'HG2133_PATCH', 'HSCHR18_1_CTG2', 'HG2554_PATCH', 'HSCHR13_1_CTG6', 'HSCHR8_7_CTG7', 'HSCHR22_1_CTG5', 'HSCHR5_3_CTG1_1', 'HG2062_PATCH', 'HG1311_HG2539_PATCH', 'HSCHR19KIR_CA04_CTG3_1', 'HSCHR19KIR_0010-5217-AB_CTG3_1', 'HG2499_PATCH', 'HSCHR22_1_CTG2', 'HSCHR2_2_CTG7_2', 'HG2442_PATCH', 'HSCHR4_2_CTG4', 'HSCHR17_3_CTG4', 'HG1395_PATCH', 'HSCHR16_2_CTG3_1', 'HSCHR19_6_CTG2', 'HSCHR17_6_CTG4', 'HG401_PATCH', 'HG2247_PATCH', 'HG2242_HG2243_PATCH', 'HG1384_PATCH', 'HSCHR5_1_CTG1', 'HSCHR21_5_CTG2', 'HSCHR6_1_CTG10', 'HSCHR16_5_CTG3_1', 'HSCHR6_8_CTG1', 'HSCHR12_5_CTG2_1', 'HSCHR6_1_CTG9', 'HSCHR21_3_CTG1_1', 'HSCHR22_2_CTG1', 'HG563_PATCH', 'HSCHR9_1_CTG2', 'HSCHR2_3_CTG1', 'HSCHR17_5_CTG4', 'HG2527_PATCH', 'HG1507_PATCH', 'HSCHR11_1_CTG3', 'HG1524_PATCH', 'HG2047_PATCH', 'HSCHR16_4_CTG1', 'HSCHR21_1_CTG1_1', 'HSCHR19_5_CTG2', 'HSCHR9_1_CTG4', 'HSCHR16_5_CTG1', 'HSCHR20_1_CTG4', 'HG1320_PATCH', 'HSCHR12_7_CTG2_1', 'HG1299_PATCH', 'HSCHR19_4_CTG2', 'HSCHR1_4_CTG32_1', 'HG1531_PATCH', 'HG1521_PATCH', 'HG460_PATCH', 'HSCHR19_2_CTG3_1', 'HSCHR12_5_CTG2', 'HG1047_PATCH', 'HSCHR13_1_CTG5', 'HSCHR8_6_CTG7', 'HG2216_PATCH', 'HG1046_PATCH', 'HG1506_PATCH', 'HG613_PATCH', 'HG1466_PATCH', 'HG1509_PATCH', 'HG2241_PATCH', 'HG494_PATCH', 'HG1817_1_PATCH', 'KI270728.1', 'KI270727.1', 'KI270442.1', 'KI270729.1', 'GL000225.1', 'KI270743.1', 'GL000008.2', 'GL000009.2', 'KI270747.1', 'KI270722.1', 'GL000194.1', 'KI270742.1', 'GL000205.2', 'GL000195.1', 'KI270736.1', 'KI270733.1', 'GL000224.1', 'GL000219.1', 'KI270719.1', 'GL000216.2', 'KI270712.1', 'KI270706.1', 'KI270725.1', 'KI270744.1', 'KI270734.1', 'GL000213.1', 'GL000220.1', 'KI270715.1', 'GL000218.1', 'KI270749.1', 'KI270741.1', 'GL000221.1', 'KI270716.1', 'KI270731.1', 'KI270751.1', 'KI270750.1', 'KI270519.1', 'GL000214.1', 'KI270708.1', 'KI270730.1', 'KI270438.1', 'KI270737.1', 'KI270721.1', 'KI270738.1', 'KI270748.1', 'KI270435.1', 'GL000208.1', 'KI270538.1', 'KI270756.1', 'KI270739.1', 'KI270757.1', 'KI270709.1', 'KI270746.1', 'KI270753.1', 'KI270589.1', 'KI270726.1', 'KI270735.1', 'KI270711.1', 'KI270745.1', 'KI270714.1', 'KI270732.1', 'KI270713.1', 'KI270754.1', 'KI270710.1', 'KI270717.1', 'KI270724.1', 'KI270720.1', 'KI270723.1', 'KI270718.1', 'KI270317.1', 'KI270740.1', 'KI270755.1', 'KI270707.1', 'KI270579.1', 'KI270752.1', 'KI270512.1', 'KI270322.1', 'GL000226.1', 'KI270311.1', 'KI270366.1', 'KI270511.1', 'KI270448.1', 'KI270521.1', 'KI270581.1', 'KI270582.1', 'KI270515.1', 'KI270588.1', 'KI270591.1', 'KI270522.1', 'KI270507.1', 'KI270590.1', 'KI270584.1', 'KI270320.1', 'KI270382.1', 'KI270468.1', 'KI270467.1', 'KI270362.1', 'KI270517.1', 'KI270593.1', 'KI270528.1', 'KI270587.1', 'KI270364.1', 'KI270371.1', 'KI270333.1', 'KI270374.1', 'KI270411.1', 'KI270414.1', 'KI270510.1', 'KI270390.1', 'KI270375.1', 'KI270420.1', 'KI270509.1', 'KI270315.1', 'KI270302.1', 'KI270518.1', 'KI270530.1', 'KI270304.1', 'KI270418.1', 'KI270424.1', 'KI270417.1', 'KI270508.1', 'KI270303.1', 'KI270381.1', 'KI270529.1', 'KI270425.1', 'KI270396.1', 'KI270363.1', 'KI270386.1', 'KI270465.1', 'KI270383.1', 'KI270384.1', 'KI270330.1', 'KI270372.1', 'KI270548.1', 'KI270580.1', 'KI270387.1', 'KI270391.1', 'KI270305.1', 'KI270373.1', 'KI270422.1', 'KI270316.1', 'KI270340.1', 'KI270338.1', 'KI270583.1', 'KI270334.1', 'KI270429.1', 'KI270393.1', 'KI270516.1', 'KI270389.1', 'KI270466.1', 'KI270388.1', 'KI270544.1', 'KI270310.1', 'KI270412.1', 'KI270395.1', 'KI270376.1', 'KI270337.1', 'KI270335.1', 'KI270378.1', 'KI270379.1', 'KI270329.1', 'KI270419.1', 'KI270336.1', 'KI270312.1', 'KI270539.1', 'KI270385.1', 'KI270423.1', 'KI270392.1', 'KI270394.1'])\n"
     ]
    }
   ],
   "source": [
    "#/home/marina/Downloads/phix.fasta\n",
    "from pyfaidx import Fasta\n",
    "\n",
    "# Path to your reference genome FASTA file\n",
    "reference_genome_file = '/mnt/raid1/thalassini/home/Downloads/Homo_sapiens.GRCh38.dna.toplevel.fa'\n",
    "\n",
    "\n",
    "\n",
    "# Load the reference genome using pyfaidx\n",
    "fasta = Fasta(reference_genome_file)\n",
    "\n",
    "# List all chromosome names in the reference genome file\n",
    "chromosome_names = fasta.keys()\n",
    "\n",
    "# Display the list of chromosome names\n",
    "print(chromosome_names)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "34faea4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace the first 25 keys with mapping_table values\n",
    "updated_fasta = {}\n",
    "\n",
    "count = 0\n",
    "for k, v in fasta.items():\n",
    "    if count < 25:\n",
    "        updated_fasta[mapping_table.get(k, k)] = v\n",
    "        count += 1\n",
    "    else:\n",
    "        updated_fasta[k] = v\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9907fc3",
   "metadata": {},
   "source": [
    "# Create true positive set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "196a351f",
   "metadata": {},
   "outputs": [],
   "source": [
    "true_positives=[]\n",
    "# Loop through the DataFrame and retrieve values of 'Chromosome', 'Start', and 'End' columns for each row\n",
    "for index, row in chipseq_filtered_df.iterrows():\n",
    "    chromosome = row['Chromosome']\n",
    "    start = row['Start']\n",
    "    end = row['End']\n",
    "    true_positives.append(updated_fasta[chromosome][start-100:end+100].seq)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "7448c169",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CGCGGCCGCCTCCTCCTCCGAACGTGGCCGCCTCCGAACGTGGCCGCCGCCTCCTCCGAACGTGGCCGCTTCCGCAGCGCCCGGCGCAGGCCGCACTCCGCCACCAGGGGGCGCCACAGCTCCTCGCGCCGCCGCCTCCCGCAAACACAAAGAGCCGCGCGGCCACGACGGCCGCGTGCCCGGAGCGCCGGGGTCTTTCCTGGGCTCCAAAGTC'"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "true_positives[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4ed72d2",
   "metadata": {},
   "source": [
    "# Create true negative set"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efa8bbf5",
   "metadata": {},
   "source": [
    "## Create negative df by excluding the rows that belong to the ensembl_filtered df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "4eec6d46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge the DataFrames based on specific columns and keep rows that do not exist in df_filtered\n",
    "merged = df.merge(filtered_df, on=['Chromosome', 'Start', 'End', 'strand', 'score', 'name'], how='outer', indicator=True)\n",
    "negative_df = merged[merged['_merge'] == 'left_only'].drop('_merge', axis=1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "67933029",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>11458</td>\n",
       "      <td>11472</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>547112</td>\n",
       "      <td>547126</td>\n",
       "      <td>-</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr4</td>\n",
       "      <td>5821787</td>\n",
       "      <td>5821801</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9495407</td>\n",
       "      <td>9495421</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr16</td>\n",
       "      <td>12095359</td>\n",
       "      <td>12095373</td>\n",
       "      <td>+</td>\n",
       "      <td>24.2951</td>\n",
       "      <td>CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451726</th>\n",
       "      <td>chr1</td>\n",
       "      <td>243056957</td>\n",
       "      <td>243056971</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_GCCTGGAGGGGCCAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451728</th>\n",
       "      <td>chr1</td>\n",
       "      <td>245560998</td>\n",
       "      <td>245561012</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_AACAGAAGGTGGCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451729</th>\n",
       "      <td>chr1</td>\n",
       "      <td>246738775</td>\n",
       "      <td>246738789</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_CCCGCGAGGTGGAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451730</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247423982</td>\n",
       "      <td>247423996</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_AGCAGGAGAGGGAGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451731</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247470114</td>\n",
       "      <td>247470128</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0984</td>\n",
       "      <td>CTCF_4.49e-05_0.548_ATCACAAGAGGTCAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>348702 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome      Start        End strand    score  \\\n",
       "0           chr10      11458      11472      +  24.2951   \n",
       "1           chr12     547112     547126      -  24.2951   \n",
       "6            chr4    5821787    5821801      +  24.2951   \n",
       "7            chr1    9495407    9495421      +  24.2951   \n",
       "10          chr16   12095359   12095373      +  24.2951   \n",
       "...           ...        ...        ...    ...      ...   \n",
       "451726       chr1  243056957  243056971      +  10.0984   \n",
       "451728       chr1  245560998  245561012      +  10.0984   \n",
       "451729       chr1  246738775  246738789      +  10.0984   \n",
       "451730       chr1  247423982  247423996      +  10.0984   \n",
       "451731       chr1  247470114  247470128      +  10.0984   \n",
       "\n",
       "                                       name  \n",
       "0       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "1       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "6       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "7       CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "10      CTCF_2.93e-10_0.026_GCCACCAGGGGGCGC  \n",
       "...                                     ...  \n",
       "451726  CTCF_4.49e-05_0.548_GCCTGGAGGGGCCAC  \n",
       "451728  CTCF_4.49e-05_0.548_AACAGAAGGTGGCCC  \n",
       "451729  CTCF_4.49e-05_0.548_CCCGCGAGGTGGAGG  \n",
       "451730  CTCF_4.49e-05_0.548_AGCAGGAGAGGGAGC  \n",
       "451731  CTCF_4.49e-05_0.548_ATCACAAGAGGTCAG  \n",
       "\n",
       "[348702 rows x 6 columns]"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "negative_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "f36fbf7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The percentage of the true negative set is 77.19222902074682% of the fimo results\n"
     ]
    }
   ],
   "source": [
    "print(f\"The percentage of the true negative set is {len(negative_df)/len(df)*100}% of the fimo results\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "14988faf",
   "metadata": {},
   "outputs": [],
   "source": [
    "true_negatives=[]\n",
    "# Loop through the DataFrame and retrieve values of 'Chromosome', 'Start', and 'End' columns for each row\n",
    "for index, row in negative_df.iterrows():\n",
    "    chromosome = row['Chromosome']\n",
    "    start = row['Start']\n",
    "    end = row['End']\n",
    "    true_negatives.append(updated_fasta[chromosome][start-100:end+100].seq)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "d50f0450",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have 348702 negative sequences and 17936 positive sequences\n"
     ]
    }
   ],
   "source": [
    "print(f\"We have {len(true_negatives)} negative sequences and {len(true_positives)} positive sequences\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "3ef482b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True positives are 4.892018830563117% of the dataset\n"
     ]
    }
   ],
   "source": [
    "print(f\"True positives are {len(true_positives)/(len(true_negatives)+len(true_positives))*100}% of the dataset\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "cae03a08",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True negatives are 95.10798116943688% of the dataset\n"
     ]
    }
   ],
   "source": [
    "print(f\"True negatives are {len(true_negatives)/(len(true_negatives)+len(true_positives))*100}% of the dataset\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aad9379a",
   "metadata": {},
   "source": [
    "## Checking the length of the sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "88970330",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the length of each string in the list\n",
    "lengths_pos = [len(item) for item in true_positives]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "5ee95499",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(214, 201)"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(lengths_pos),min(lengths_pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "79ea671e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the length of each string in the list\n",
    "lengths_neg = [len(item) for item in true_negatives]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "300c84d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "max(lengths_neg),min(lengths_neg)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc30e9e5",
   "metadata": {},
   "source": [
    "# Combine the two sets in a df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "35af65da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                     data  class\n",
      "0       GCCCCTGGGCCACGCCCACTCCCCCAAGCGGGGAAGGAGCTTCGCG...      0\n",
      "1       CACCCACAACAGCCTGGACCCCGGACCGAGGTCCCCCTCGGTCGTC...      0\n",
      "2       AAAACACTGACAGGAAAAGGGATGGACATGATTCCCAGAATCCTTC...      0\n",
      "3       TAGTAGGTCATACCGCAGTGCCCACCGCAAGGGTGGCCACTGCGGA...      0\n",
      "4       AGTTAATTTCGTTTGAATATATGGTTATTTGGGAGCTGGAGGTTGC...      0\n",
      "...                                                   ...    ...\n",
      "366633  CGGCCTATGACGGCGTCCTTGCGGAAGTACGGGTGCCGAGAGAAGC...      1\n",
      "366634  TTGATCTTGTTTGTTACCTGGTTCTCAGTGAGCATAGGTAGTAGAC...      1\n",
      "366635  TTGATCTTGTTTGTTACCTGGTTCTCAGTGAGCATAGGTAGTAGAC...      1\n",
      "366636  CACCCGTTTCTATATTGAGGAGCTCCTCTCCATAAATAAAGCACAG...      1\n",
      "366637  GGGGCGGTGGCGCGGAGTGGGCGGTGACTTCGGCGGGCGCCTCCCA...      1\n",
      "\n",
      "[366638 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Create DataFrames for negative and positive sequences\n",
    "true_negatives_df = pd.DataFrame({'data': true_negatives, 'class': 0})\n",
    "true_positives_df= pd.DataFrame({'data': true_positives, 'class': 1})\n",
    "\n",
    "# Concatenate both DataFrames into a single DataFrame\n",
    "combined_df = pd.concat([true_negatives_df, true_positives_df], ignore_index=True)\n",
    "\n",
    "# Display the combined DataFrame\n",
    "print(combined_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "0684c8e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the DataFrame to a CSV file\n",
    "combined_df.to_csv('data.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "18ed0f4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Shuffle the rows of the DataFrame\n",
    "combined_df = combined_df.sample(frac=1, random_state=42).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "756890e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from tensorflow.keras.preprocessing.text import Tokenizer\n",
    "from tensorflow.keras.preprocessing.sequence import pad_sequences\n",
    "from tensorflow.keras.models import Sequential\n",
    "from tensorflow.keras.layers import Embedding, Conv1D, MaxPooling1D, Flatten, Dense\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Tokenizing sequences and converting to sequences of integers\n",
    "tokenizer = Tokenizer(char_level=True)\n",
    "tokenizer.fit_on_texts(combined_df['data'])\n",
    "sequences = tokenizer.texts_to_sequences(combined_df['data'])\n",
    "\n",
    "\n",
    "\n",
    "# Perform one-hot encoding for the padded sequences\n",
    "#one_hot_sequences = np.array([one_hot_encode(seq) for seq in combined_df['data']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ad71a43",
   "metadata": {},
   "outputs": [],
   "source": [
    "def one_hot_encode(sequence):\n",
    "    encoding = {'A': [1, 0, 0, 0], 'T': [0, 1, 0, 0], 'G': [0, 0, 1, 0], 'C': [0, 0, 0, 1]}\n",
    "    one_hot_sequence = [encoding.get(base, [0, 0, 0, 0]) for base in sequence]\n",
    "    return np.array(one_hot_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "4d2ffc87",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find the maximum length of sequences\n",
    "max_length = max(len(seq) for seq in combined_df['data'])\n",
    "\n",
    "# Pad sequences to ensure they all have the same length\n",
    "padded_sequences = [seq.ljust(max_length, 'N') for seq in combined_df['data']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "8d17ed13",
   "metadata": {},
   "outputs": [],
   "source": [
    "X= [one_hot_encode(sequence) for sequence in padded_sequences]\n",
    "X=np.array(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "6b729e00",
   "metadata": {},
   "outputs": [],
   "source": [
    "y=np.array(combined_df['class'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "df870b58",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(tensor([[[0., 0., 1., 0.],\n",
       "          [1., 0., 0., 0.],\n",
       "          [0., 1., 0., 0.],\n",
       "          ...,\n",
       "          [0., 1., 0., 0.],\n",
       "          [0., 0., 0., 1.],\n",
       "          [0., 0., 0., 1.]],\n",
       " \n",
       "         [[0., 1., 0., 0.],\n",
       "          [0., 0., 1., 0.],\n",
       "          [0., 0., 1., 0.],\n",
       "          ...,\n",
       "          [0., 1., 0., 0.],\n",
       "          [0., 1., 0., 0.],\n",
       "          [0., 0., 1., 0.]],\n",
       " \n",
       "         [[0., 0., 0., 1.],\n",
       "          [0., 0., 0., 1.],\n",
       "          [0., 0., 0., 1.],\n",
       "          ...,\n",
       "          [0., 1., 0., 0.],\n",
       "          [0., 0., 1., 0.],\n",
       "          [0., 0., 1., 0.]],\n",
       " \n",
       "         [[0., 1., 0., 0.],\n",
       "          [0., 0., 1., 0.],\n",
       "          [0., 0., 1., 0.],\n",
       "          ...,\n",
       "          [0., 0., 1., 0.],\n",
       "          [0., 0., 1., 0.],\n",
       "          [0., 1., 0., 0.]],\n",
       " \n",
       "         [[0., 0., 0., 1.],\n",
       "          [0., 1., 0., 0.],\n",
       "          [0., 0., 1., 0.],\n",
       "          ...,\n",
       "          [0., 0., 1., 0.],\n",
       "          [1., 0., 0., 0.],\n",
       "          [0., 0., 1., 0.]]]),\n",
       " tensor([0., 0., 0., 0., 0.]))"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import torch\n",
    "X = torch.from_numpy(X).type(torch.float)\n",
    "y = torch.from_numpy(y).type(torch.float)\n",
    "\n",
    "# View the first five samples\n",
    "X[:5], y[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "2195c240",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(293310, 73328, 293310, 73328)"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Split data into train and test sets\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, \n",
    "                                                    y, \n",
    "                                                    test_size=0.2, # 20% test, 80% train\n",
    "                                                    random_state=42) # make the random split reproducible\n",
    "\n",
    "len(X_train), len(X_test), len(y_train), len(y_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ffb1b51",
   "metadata": {},
   "source": [
    "# Set up device agnostic code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "b9e55dd9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'cpu'"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Standard PyTorch imports\n",
    "import torch\n",
    "from torch import nn\n",
    "\n",
    "# Make device agnostic code\n",
    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
    "device"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "3a72491b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ModelV0(\n",
       "  (layer_1): Linear(in_features=4, out_features=5, bias=True)\n",
       "  (layer_2): Linear(in_features=5, out_features=1, bias=True)\n",
       ")"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 1. Construct a model class that subclasses nn.Module\n",
    "class ModelV0(nn.Module):\n",
    "    def __init__(self):\n",
    "        super().__init__()\n",
    "        # 2. Create 2 nn.Linear layers capable of handling X and y input and output shapes\n",
    "        self.layer_1 = nn.Linear(in_features=4, out_features=5) # takes in 2 features (X), produces 5 features\n",
    "        self.layer_2 = nn.Linear(in_features=5, out_features=1) # takes in 5 features, produces 1 feature (y)\n",
    "    \n",
    "    # 3. Define a forward method containing the forward pass computation\n",
    "    def forward(self, x):\n",
    "        # Return the output of layer_2, a single feature, the same shape as y\n",
    "        return self.layer_2(self.layer_1(x)) # computation goes through layer_1 first then the output of layer_1 goes through layer_2\n",
    "\n",
    "# 4. Create an instance of the model and send it to target device\n",
    "model_0 = ModelV0().to(device)\n",
    "model_0\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b5cb06f",
   "metadata": {},
   "source": [
    "# Set up loss function and optimizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "id": "d3290451",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a loss function\n",
    "# loss_fn = nn.BCELoss() # BCELoss = no sigmoid built-in\n",
    "loss_fn = nn.BCEWithLogitsLoss() # BCEWithLogitsLoss = sigmoid built-in\n",
    "\n",
    "# Create an optimizer\n",
    "optimizer = torch.optim.SGD(params=model_0.parameters(), \n",
    "                            lr=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "f06051fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate accuracy (a classification metric)\n",
    "def accuracy_fn(y_true, y_pred):\n",
    "    correct = torch.eq(y_true, y_pred).sum().item() # torch.eq() calculates where two tensors are equal\n",
    "    acc = (correct / len(y_pred)) * 100 \n",
    "    return acc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "349db31b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import balanced_accuracy_score\n",
    "\n",
    "# Function to calculate balanced accuracy\n",
    "def balanced_accuracy(y_true, y_pred):\n",
    "    return balanced_accuracy_score(y_true, y_pred) * 100"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f91605c1",
   "metadata": {},
   "source": [
    "# Train the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "b5c88056",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([[[-0.1812],\n",
       "         [-0.4116],\n",
       "         [-0.4116],\n",
       "         ...,\n",
       "         [-0.5166],\n",
       "         [-0.5166],\n",
       "         [-0.0459]],\n",
       "\n",
       "        [[-0.4116],\n",
       "         [-0.5166],\n",
       "         [-0.4116],\n",
       "         ...,\n",
       "         [-0.4116],\n",
       "         [-0.4116],\n",
       "         [-0.4116]],\n",
       "\n",
       "        [[-0.1812],\n",
       "         [-0.4116],\n",
       "         [-0.4116],\n",
       "         ...,\n",
       "         [-0.0459],\n",
       "         [-0.4116],\n",
       "         [-0.0459]],\n",
       "\n",
       "        [[-0.0459],\n",
       "         [-0.1812],\n",
       "         [-0.5166],\n",
       "         ...,\n",
       "         [-0.0459],\n",
       "         [-0.4116],\n",
       "         [-0.4116]],\n",
       "\n",
       "        [[-0.1812],\n",
       "         [-0.1812],\n",
       "         [-0.1812],\n",
       "         ...,\n",
       "         [-0.1812],\n",
       "         [-0.5166],\n",
       "         [-0.1812]]], grad_fn=<SliceBackward0>)"
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# View the frist 5 outputs of the forward pass on the test data\n",
    "y_logits = model_0(X_test.to(device))[:5]\n",
    "y_logits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "3c08391e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([[[0.4548],\n",
       "         [0.3985],\n",
       "         [0.3985],\n",
       "         ...,\n",
       "         [0.3736],\n",
       "         [0.3736],\n",
       "         [0.4885]],\n",
       "\n",
       "        [[0.3985],\n",
       "         [0.3736],\n",
       "         [0.3985],\n",
       "         ...,\n",
       "         [0.3985],\n",
       "         [0.3985],\n",
       "         [0.3985]],\n",
       "\n",
       "        [[0.4548],\n",
       "         [0.3985],\n",
       "         [0.3985],\n",
       "         ...,\n",
       "         [0.4885],\n",
       "         [0.3985],\n",
       "         [0.4885]],\n",
       "\n",
       "        [[0.4885],\n",
       "         [0.4548],\n",
       "         [0.3736],\n",
       "         ...,\n",
       "         [0.4885],\n",
       "         [0.3985],\n",
       "         [0.3985]],\n",
       "\n",
       "        [[0.4548],\n",
       "         [0.4548],\n",
       "         [0.4548],\n",
       "         ...,\n",
       "         [0.4548],\n",
       "         [0.3736],\n",
       "         [0.4548]]], grad_fn=<SigmoidBackward0>)"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Use sigmoid on model logits\n",
    "y_pred_probs = torch.sigmoid(y_logits)\n",
    "y_pred_probs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "db8e4ec3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tensor([[True, True, True,  ..., True, True, True],\n",
      "        [True, True, True,  ..., True, True, True],\n",
      "        [True, True, True,  ..., True, True, True],\n",
      "        [True, True, True,  ..., True, True, True],\n",
      "        [True, True, True,  ..., True, True, True]])\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "tensor([[0., 0., 0.,  ..., 0., 0., 0.],\n",
       "        [0., 0., 0.,  ..., 0., 0., 0.],\n",
       "        [0., 0., 0.,  ..., 0., 0., 0.],\n",
       "        [0., 0., 0.,  ..., 0., 0., 0.],\n",
       "        [0., 0., 0.,  ..., 0., 0., 0.]], grad_fn=<SqueezeBackward0>)"
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find the predicted labels (round the prediction probabilities)\n",
    "y_preds = torch.round(y_pred_probs)\n",
    "\n",
    "# In full\n",
    "y_pred_labels = torch.round(torch.sigmoid(model_0(X_test.to(device))[:5]))\n",
    "\n",
    "# Check for equality\n",
    "print(torch.eq(y_preds.squeeze(), y_pred_labels.squeeze()))\n",
    "\n",
    "# Get rid of extra dimension\n",
    "y_preds.squeeze()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "7f036f68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([1., 0., 0., 0., 0.])"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_test[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "233d5d62",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
